src/Agent.RepositoryPlugin/RepositoryPlugin.cs

```diff
@@ -0,0 +1,136 @@
+    {
+        Task GetSourceAsync(AgentTaskPluginExecutionContext executionContext, Pipelines.RepositoryResource repository, CancellationToken cancellationToken);
+
+        Task PostJobCleanupAsync(AgentTaskPluginExecutionContext executionContext, Pipelines.RepositoryResource repository);
+    }
+
+    public class CheckoutTask : IAgentTaskPlugin
+    {
+        public string FriendlyName => "Get Sources";
+        public Guid Id => new Guid("c61807ba-5e20-4b70-bd8c-3683c9f74003");
+        public string Version => "1.0.0";
+        public string Description => "Get Sources";
+        public string HelpMarkDown => "";
+        public string Author => "Microsoft";
+
+        public TaskInputDefinition[] Inputs => new TaskInputDefinition[] {
+            new TaskInputDefinition()
+            {
+                Name="repository",
+                InputType = TaskInputType.String,
+                DefaultValue="self",
+                Required=true
+            }
+        };
+
+        public HashSet<string> Stages => new HashSet<string>() { "main", "post" };
+
+        public async Task RunAsync(AgentTaskPluginExecutionContext executionContext, CancellationToken token)
+        {
+            var repoAlias = executionContext.GetInput("repository", true);
+            var repo = executionContext.Repositories.Single(x => string.Equals(x.Alias, repoAlias, StringComparison.OrdinalIgnoreCase));
+            MergeInputs(executionContext, repo);
+
+            ISourceProvider sourceProvider = null;
+            switch (repo.Type)
+            {
+                case RepositoryTypes.Bitbucket:
+                case RepositoryTypes.GitHub:
+                case RepositoryTypes.GitHubEnterprise:
+                    sourceProvider = new AuthenticatedGitSourceProvider();
+                    break;
+                case RepositoryTypes.Git:
+                    sourceProvider = new ExternalGitSourceProvider();
+                    break;
+                case RepositoryTypes.TfsGit:
+                    sourceProvider = new TfsGitSourceProvider();
+                    break;
+                case RepositoryTypes.TfsVersionControl:
+                    sourceProvider = new TfsVCSourceProvider();
+                    break;
+                case RepositoryTypes.Svn:
+                    sourceProvider = new SvnSourceProvider();
+                    break;
+                default:
+                    throw new NotSupportedException(repo.Type);
+            }
+
+            if (executionContext.Stage == "main")
+            {
+                await sourceProvider.GetSourceAsync(executionContext, repo, token);
+            }
+            else if (executionContext.Stage == "post")
+            {
+                await sourceProvider.PostJobCleanupAsync(executionContext, repo);
+            }
+        }
+
+        private void MergeInputs(AgentTaskPluginExecutionContext executionContext, Pipelines.RepositoryResource repository)
```

Comment:
lets discuss offline. i'm confused why this is here

Reply:
i will deal with this later when we send repository as part of the message.